arch/arm/src/nrf91/nrf91_errata.c

```diff
@@ -0,0 +1,193 @@
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define MEMORY_SYNC() ARM_ISB(); ARM_DSB()
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+#ifdef CONFIG_NRF91_UICR_HFXO_WORKAROUND
+/****************************************************************************
+ * Name: nrf91_uicr_recover
+ *
+ * Description:
+ *   Recover HFXOCNT and HFXOSRC registers after erasing all FLASH.
+ *   Without this workaround, the modem core never boots properly.
+ *
+ ****************************************************************************/
+
+void nrf91_uicr_recover(void)
+{
+  bool hfxosrc_recover = false;
+  bool hfxocnt_recover = false;
+
+  if (getreg32(NRF91_UICR_HFXOCNT) == 0xffffffff)
+    {
+      hfxocnt_recover = true;
+    }
+
+  if (getreg32(NRF91_UICR_HFXOSRC) == 0xffffffff)
+    {
+      hfxosrc_recover = true;
+    }
+
+  if (!hfxocnt_recover && !hfxosrc_recover)
+    {
+      return;
+    }
+
+  /* Wait for flash */
+
+  while (!(getreg32(NRF91_NVMC_READY) & NVMC_READY_READY))
+    {
+    }
+
+  /* Enable write */
+
+  putreg32(NVMC_CONFIG_WEN, NRF91_NVMC_CONFIG);
+
+  /* Memory sync */
+
+  MEMORY_SYNC();
+
+  /* Write HFXOSRC */
+
+  if (hfxosrc_recover)
+    {
+      putreg32(0x0 , NRF91_UICR_HFXOSRC);
+    }
+
+  /* Wait for flash */
+
+  while (!(getreg32(NRF91_NVMC_READY) & NVMC_READY_READY))
+    {
+    }
+
+  /* Memory sync */
+
+  MEMORY_SYNC();
+
+  /* Write HFXOCNT */
+
+  if (hfxocnt_recover)
+    {
+      putreg32(0x20, NRF91_UICR_HFXOCNT);
+    }
+
+  /* Wait for flash */
+
+  while (!(getreg32(NRF91_NVMC_READY) & NVMC_READY_READY))
+    {
+    }
+
+  /* Memory sync */
+
+  MEMORY_SYNC();
+
+  /* Read only access */
+
+  putreg32(NVMC_CONFIG_REN, NRF91_NVMC_CONFIG);
+
+  /* System reset */
+
+  up_systemreset();
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+#ifndef CONFIG_ARCH_TRUSTZONE_NONSECURE
+/****************************************************************************
+ * Name: nrf91_errata_secure
+ *
+ * Description:
+ *   Errata that must be applied in asecure environment
+ *
+ ****************************************************************************/
+
+void nrf91_errata_secure(void)
+{
+#ifdef NRF91_DCDC_ERRATA15_WORKAROUND
+  /* Workaround for 3.7 [15] REGULATORS: Supply regulators default to
+   * LDO mode after reset
+   */
+
+  putreg32(REGULATORS_DCDCEN_ENABLE, NRF91_REGULATORS_DCDCEN);
+#endif
+
+#ifdef NRF91_LFXO_ERRATA31_WORKAROUND
+  /* Workaround for 3.15 [31] LFXO: LFXO startup fails */
+
+  *((volatile uint32_t *)0x5000470c) = 0x0;
+  *((volatile uint32_t *)0x50004710) = 0x1;
```

Comment:
Why not `putreg32`?

Reply:
To keep it compatible with the errata from Nordic. They use this format to define a workaround.

Reply:
Thanks for explanation.